src/System/Remote/Monitoring/CloudWatch.hs

```diff
@@ -141,36 +149,57 @@ diffSamples prev !curr = Map.foldlWithKey' combine Map.empty curr
         d2 {Distribution.count = Distribution.count d2 - Distribution.count d1}
     diffMetric _ _ = Nothing
 
-flushSample :: CloudWatchEnv -> Metrics.Sample -> IO ()
-flushSample CloudWatchEnv{..} = void . Map.traverseWithKey flushMetric
+metricToDatum :: [Dimension] -> Text -> Metrics.Value -> Maybe MetricDatum
+metricToDatum dim name =
+  \case
```

Comment:
I don't think I've ever known anyone that loves LambdaCase as much as you do. I hope it doesn't affect CAF'ing like point-free does.

Reply:
to be clear, I get it. You don't want to re-declare `dim` and `name` for each case.

Reply:
You'll have to blame that one on Ben :) He got me into it.

It would be easy to refactor it to the `case datum of ...` format. Knowing the CAF implication would be interesting too.

Reply:
>Knowing the CAF implication would be interesting too.

Basically you should avoid point-free top level declarations because GHC'll treats syntactic values that are typed like functions as if they are values, not functions. It'll float them out and CAF them, leading to GHC gobbling up heap and holding onto it like Smaug protecting his gold hoard.

Briefly: never ever write top-level point-free declarations. CAF-ination hinges on syntax and not types.

FWIW, I don't think LambdaCase trips this but I'm hyper-paranoid about it lifting out and CAF'ing for the final argument.

Reply:
I'd be surprised if it didn't trip this. My understanding is that lambdacase desugars to

```haskell
\case
  {pats}
-- to
\x -> case x of
  {pats}
```

"No point-free top level declarations" is not something I knew. Thanks for the tip~ Do you know any good reading on that?

Reply:
>"No point-free top level declarations" is not something I knew. Thanks for the tip~ Do you know any good reading on that?

Just experimentation that arose from superstitious rule transmission from Carter.

https://ghc.haskell.org/trac/ghc/ticket/917

https://stackoverflow.com/a/6094427/1205702

Semi-related: https://stackoverflow.com/questions/38428512/why-does-the-pointfree-version-of-my-function-use-much-more-memory

Reply:
this deserves a blog post!

Reply:
I'll think about it. I'd need a repro and that sort of thing can be fiddly with GHC.